modules/mcc/include/opencv2/mcc/ccm.hpp

```diff
@@ -0,0 +1,351 @@
+            type: double;
+            default: 0;
+    ---------------------------------------------------
+    initial_method_type :
+            the method of calculating CCM initial value;
+            see CCM.pdf for details;
+            Supported list:
+                'LEAST_SQUARE': least-squre method;
+                'WHITE_BALANCE': white balance method;
+            type: enum INITIAL_METHOD_TYPE;
+    max_count, epsilon :
+            used in MinProblemSolver-DownhillSolver;
+            Terminal criteria to the algorithm;
+            type: int, double;
+            default: 5000, 1e-4;
+    ---------------------------------------------------
+    Supported Color Space:
+            Supported list of RGB color spaces:
+                sRGB;
+                AdobeRGB;
+                WideGamutRGB;
+                ProPhotoRGB;
+                DCI_P3_RGB;
+                AppleRGB;
+                REC_709_RGB;
+                REC_2020_RGB;
+            Supported list of linear RGB color spaces:
+                sRGBL;
+                AdobeRGBL;
+                WideGamutRGBL;
+                ProPhotoRGBL;
+                DCI_P3_RGBL;
+                AppleRGBL;
+                REC_709_RGBL;
+                REC_2020_RGBL;
+            Supported list of non-RGB color spaces:
+                Lab_D50_2;
+                Lab_D65_2;
+                XYZ_D50_2;
+                XYZ_D65_2;
+            Supported IO (You can use Lab(io) or XYZ(io) to create color space):
+                A_2;
+                A_10;
+                D50_2;
+                D50_10;
+                D55_2;
+                D55_10;
+                D65_2;
+                D65_10;
+                D75_2;
+                D75_10;
+                E_2;
+                E_10;
+    ---------------------------------------------------
+    Abbr.
+        src, s: source;
+        dst, d: destination;
+        io: illuminant & observer;
+        sio, dio: source of io; destination of io;
+        rgbl: linear RGB
+        cs: color space;
+        cc: Colorchecker;
+        M, m: matrix
+        ccm: color correction matrix;
+        cam: chromatic adaption matrix;
+*/
+
+
+/* *\ brief Enum of the possible types of ccm.
+*/
+enum CCM_TYPE
+{
+    CCM_3x3,
+    CCM_4x3
+};
+
+/* *\ brief Enum of the possible types of initial method.
+*/
+enum INITIAL_METHOD_TYPE
+{
+    WHITE_BALANCE,
+    LEAST_SQUARE
+};
+
+
+/* *\ brief Core class of ccm model.
+    *        produce a ColorCorrectionModel instance for inference.
+*/
+
+class CV_EXPORTS_W ColorCorrectionModel
+{
+public:
+    // detected colors, the referenceand the RGB colorspace for conversion
+    Mat src;
+    Color dst;
+    Mat dist;
+    RGBBase_& cs;
+    Mat mask;
+
+    // RGBl of detected data and the reference
+    Mat src_rgbl;
+    Mat dst_rgbl;
+
+    // ccm type and shape
+    CCM_TYPE ccm_type;
+    int shape;
+
+    // linear method and distance
+    std::shared_ptr<Linear> linear;
```

Comment:
@alalek To make the API more simple, do you mean using a normal pointer instead of a smart pointer? Or could you tell us some advice or example to fix the issue？

Reply:
I think he means forward declaration of structure or class (they are the same thing in cpp) that is then implemented in a .cpp file, and then point to that forward declared type, see:
https://en.cppreference.com/w/cpp/language/pimpl 
where they forward declare a structure "impl" and then point to it. The implementation is in another file. 

Also, should we make unnecessary variables for the user private?

@alalek Is this correct?

Reply:
> do you mean using a normal pointer instead of a smart pointer

It is not about `linear` field only. This comment covers other fields too.

Need to hide all implementation-specific fields in .cpp file.
See `QRCodeDetector` and `QRCodeDetector::Impl` in objdetect.hpp as example.

Reply:
Thank you for your detailed explanation, I got it.